src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -68,37 +63,21 @@ def run_python_test(test_target, pytest, python_setup, source_root_config, subpr
   # https://pytest.org/en/latest/goodpractices.html#test-discovery. In addition to a performance
   # optimization, this ensures that any transitive sources, such as a test project file named
   # test_fail.py, do not unintentionally end up being run as tests.
-  source_roots = source_root_config.get_source_roots()
-  test_target_sources_file_names = []
-  for source_target_filename in test_target.sources.files_relative_to_buildroot:
-    source_root = source_roots.find_by_path(source_target_filename)
-    test_target_sources_file_names.append(
-      strip_prefix(source_target_filename, prefix=f"{source_root.path}/")
-    )
+  source_root_stripped_test_target_sources = yield Get(
+    SourceRootStrippedSources, TargetAdaptor, test_target
+  )
+  test_target_sources_file_names = sorted(source_root_stripped_test_target_sources.snapshot.files)
 
-  # Gather sources and adjust for source roots.
-  # TODO: make TargetAdaptor return a 'sources' field with an empty snapshot instead of raising to
-  # simplify the hasattr() checks here!
-  sources_digest_to_source_roots: Dict[Digest, Optional[SourceRoot]] = {}
-  for maybe_source_target in all_targets:
-    if not hasattr(maybe_source_target, 'sources'):
-      continue
-    digest = maybe_source_target.sources.snapshot.directory_digest
-    source_root = source_roots.find_by_path(maybe_source_target.address.spec_path)
-    if maybe_source_target.type_alias == Files.alias():
-      # Loose `Files`, as opposed to `Resources` or `PythonTarget`s, have no (implied) package
-      # structure and so we do not remove their source root like we normally do, so that Python
-      # filesystem APIs may still access the files. See pex_build_util.py's `_create_source_dumper`.
-      source_root = None
-    sources_digest_to_source_roots[digest] = source_root.path if source_root else ""
-
-  stripped_sources_digests = yield [
-    Get(Digest, DirectoryWithPrefixToStrip(directory_digest=digest, prefix=source_root))
-    for digest, source_root in sources_digest_to_source_roots.items()
+  source_root_stripped_sources = yield [
+    Get(SourceRootStrippedSources, TargetAdaptor, target_adaptor)
```

Comment:
So, as discussed offline, the reason this fails is because `target_adaptor` is actually "some unknown subclass of `TargetAdaptor`", and the engine does not support inheritance.

I referenced #4535, and it is indeed very relevant.

But note the way that you were able to get this `all_targets` list in the first place: by consuming a wrapper type called `HydratedTarget` that holds "`TargetAdaptor` subclasses". So in this case, the solution might be to just hold-onto and then project the relevant `HydratedTarget` instances here for consumption by your other `@rule`.